### PR TITLE
clean up IR function registry code

### PR DESCRIPTION
**Pull request by catoverdrive**
use MultiMap instead of `Map[String, Seq[]]` and pull out duplicate code between `lookupFunction` and `lookupConversion`.

### CODE REVIEW

File: src/main/scala/is/hail/expr/ir/functions/Functions.scala [hunk 1]
```diff
@@ -11,64 +11,54 @@ import scala.collection.mutable
...
+    reg.lift(name).flatMap { fs =>
+      fs.filter { case (ts, _) =>
+        ts.length == args.length && {
+          ts.foreach(_.clear())
+          (ts, args).zipped.forall(_.unify(_))
+        }
+      }.toSeq match {
+        case Seq() => None
+        case Seq((_, f)) => Some(f)
+        case _ => fatal(s"Multiple functions found that satisfy $name$args.")
```
Comment (jigold) [inline src/main/scala/is/hail/expr/ir/functions/Functions.scala]: can you make this error message more readable. `$name(${ args.mkString(",") })`

File: src/main/scala/is/hail/expr/ir/functions/Functions.scala [hunk 2]
```diff
@@ -11,64 +11,54 @@ import scala.collection.mutable
...
     }
 
-    validMethods match {
-      case Seq() => None
-      case Seq(x) => Some(x)
+    (validIR, validMethods) match {
+      case (None, None) => None
+      case (None, Some(x)) => Some(x)
+      case (Some(x), None) => Some(x)
       case _ => fatal(s"Multiple methods found that satisfy $name$args.")
```
Comment (jigold) [inline src/main/scala/is/hail/expr/ir/functions/Functions.scala]: error message - see note above.

File: src/main/scala/is/hail/expr/ir/functions/Functions.scala [hunk 3]
```diff
@@ -11,64 +11,54 @@ import scala.collection.mutable
...
-          Some(f)
-        else
-          None
-      } else
-        None
-    }
 
-    val validMethods = validIR ++ lookupFunction(name, args).map { f =>
-      { args: Seq[IR] =>
+    val validIR = lookupInRegistry(irRegistry, name, args)
```
Comment (jigold) [inline src/main/scala/is/hail/expr/ir/functions/Functions.scala]: Might be nice to have the wrapper function `lookupIR` analogous to `lookupFunction`. Not wedded to this though.

**Comment by jigold** [comment]
Could you please fix the error messages? See above.

**Comment by catoverdrive** [comment]
oh, I totally forgot to push the commit. womp.

**Comment by catoverdrive** [comment]
(done)

**Pull request merged**